Test container exiting message for tcp readiness probe

yt3liu opened the pull request:
<!--
Request Prow to automatically lint any go code in this PR:

/lint
-->

Adding tcp method coverage for readiness probe. 

Part of #3486

mattmoor commented:
/retest

wtam2018 commented on test/conformance/errorcondition_test.go:
```diff
@@ -118,12 +118,6 @@ func TestContainerErrorMsg(t *testing.T) {
 // for the container crashing scenario.
 func TestContainerExitingMsg(t *testing.T) {
 	t.Parallel()
```
I believe you can delete `t.Parallel()` here as it is called by each test later.

yt3liu replied:
Deleted.

dgerd replied:
The original `t.Parallel()` does change the behavior of this test. A `t.Parallel()` in subtests will allow subtests to run in parallel, but will not allow the overall test to run parallel with other conformance tests.

See about halfway down https://blog.golang.org/subtests for details.

yt3liu replied:
Thanks for the reference. Added back the original `t.Parallel()` to allow this group of tests to be run in parallel along with the other conformance test.

wtam2018 commented on test/conformance/errorcondition_test.go:
```diff
@@ -132,65 +126,90 @@ func TestContainerExitingMsg(t *testing.T) {
+			test.CleanupOnInterrupt(func() { test.TearDown(clients, names) })
+
+			t.Log("When the containers keep crashing, the Configuration should have error status.")
+
```
I'd include the test name (http vs tcp) in the logging as they are running n parallel.

yt3liu replied:
The output does not get printed until the test is completed. Thus, the log messages are not mixed up. From the current output, I think it's clear which test is run, but feel free to disagree and suggest other improvements. 

Here's what a sample output: 
```
=== RUN   TestContainerExitingMsg
=== RUN   TestContainerExitingMsg/http
=== PAUSE TestContainerExitingMsg/http
=== RUN   TestContainerExitingMsg/tcp
=== PAUSE TestContainerExitingMsg/tcp
=== CONT  TestContainerExitingMsg/http
=== CONT  TestContainerExitingMsg/tcp
--- PASS: TestContainerExitingMsg (0.00s)
    --- PASS: TestContainerExitingMsg/tcp (121.49s)
        errorcondition_test.go:157: Creating a new Configuration failing
        util.go:34: resource (test.ResourceObjects) {
             Route: (*v1alpha1.Route)(<nil>),
             Config: (*v1alpha1.Configuration)(0xc0001d4400)({
              TypeMeta: (v1.TypeMeta) &TypeMeta{Kind:,APIVersion:,},
              ObjectMeta: (v1.ObjectMeta) &ObjectMeta{Name:container-exiting-msg-tcp-rlxhqmkq,GenerateName:,Namespace:serving-tests,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,},
              Spec: (v1alpha1.ConfigurationSpec) {
               DeprecatedGeneration: (int64) 0,
               Build: (*v1alpha1.RawExtension)(<nil>),
               RevisionTemplate: (v1alpha1.RevisionTemplateSpec) &ObjectMeta{Name:,GenerateName:,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,}
              },
              Status: (v1alpha1.ConfigurationStatus) {
               Status: (v1alpha1.Status) {
                ObservedGeneration: (int64) 0,
                Conditions: (v1alpha1.Conditions) <nil>
               },
               ConfigurationStatusFields: (v1alpha1.ConfigurationStatusFields) {
                LatestReadyRevisionName: (string) "",
                LatestCreatedRevisionName: (string) ""
               }
              }
             }),
             Service: (*v1alpha1.Service)(<nil>),
             Revision: (*v1alpha1.Revision)(<nil>)
            }
        errorcondition_test.go:165: When the containers keep crashing, the Configuration should have error status.
        errorcondition_test.go:189: When the containers keep crashing, the revision should have error status.
        errorcondition_test.go:206: When the revision has error condition, logUrl should be populated.
    --- PASS: TestContainerExitingMsg/http (122.53s)
        errorcondition_test.go:157: Creating a new Configuration failing
        util.go:34: resource (test.ResourceObjects) {
             Route: (*v1alpha1.Route)(<nil>),
             Config: (*v1alpha1.Configuration)(0xc00052e400)({
              TypeMeta: (v1.TypeMeta) &TypeMeta{Kind:,APIVersion:,},
              ObjectMeta: (v1.ObjectMeta) &ObjectMeta{Name:container-exiting-msg-http-gveihupq,GenerateName:,Namespace:serving-tests,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,},
              Spec: (v1alpha1.ConfigurationSpec) {
               DeprecatedGeneration: (int64) 0,
               Build: (*v1alpha1.RawExtension)(<nil>),
               RevisionTemplate: (v1alpha1.RevisionTemplateSpec) &ObjectMeta{Name:,GenerateName:,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,}
              },
              Status: (v1alpha1.ConfigurationStatus) {
               Status: (v1alpha1.Status) {
                ObservedGeneration: (int64) 0,
                Conditions: (v1alpha1.Conditions) <nil>
               },
               ConfigurationStatusFields: (v1alpha1.ConfigurationStatusFields) {
                LatestReadyRevisionName: (string) "",
                LatestCreatedRevisionName: (string) ""
               }
              }
             }),
             Service: (*v1alpha1.Service)(<nil>),
             Revision: (*v1alpha1.Revision)(<nil>)
            }
        errorcondition_test.go:165: When the containers keep crashing, the Configuration should have error status.
        errorcondition_test.go:189: When the containers keep crashing, the revision should have error status.
        errorcondition_test.go:206: When the revision has error condition, logUrl should be populated.
PASS
ok  	github.com/knative/serving/test/conformance	122.948s

```

wtam2018 replied:
Yes.  They all look good when they pass.   However, when something goes wrong, we may need to match the logs to figure out which message has failed/gone missing.   It probably is worth logging the extra info.   WDYT?

yt3liu replied:
Adding a sample output for a failed test. For http, I made it always fail on validation configuration state [Line 180](https://github.com/knative/serving/blob/77abfd961a6a1fb7116e915ac24bf7fea0e7363e/test/conformance/errorcondition_test.go#L180) by checking the test name. Similarly, made tcp to fail on validation on the revision state [Line 202](https://github.com/knative/serving/blob/77abfd961a6a1fb7116e915ac24bf7fea0e7363e/test/conformance/errorcondition_test.go#L202). 

In my opinion, I think the current logging is good enough. Is there a particular failure case where you think it would mix up the log/error messages? 

<details>
  <summary>Sample failed output</summary>
  
  ```
=== RUN   TestContainerExitingMsg
=== RUN   TestContainerExitingMsg/http
=== PAUSE TestContainerExitingMsg/http
=== RUN   TestContainerExitingMsg/tcp
=== PAUSE TestContainerExitingMsg/tcp
=== CONT  TestContainerExitingMsg/http
=== CONT  TestContainerExitingMsg/tcp
--- FAIL: TestContainerExitingMsg (0.00s)
    --- FAIL: TestContainerExitingMsg/http (18.22s)
        errorcondition_test.go:157: Creating a new Configuration failing
        util.go:34: resource (test.ResourceObjects) {
             Route: (*v1alpha1.Route)(<nil>),
             Config: (*v1alpha1.Configuration)(0xc00058a400)({
              TypeMeta: (v1.TypeMeta) &TypeMeta{Kind:,APIVersion:,},
              ObjectMeta: (v1.ObjectMeta) &ObjectMeta{Name:container-exiting-msg-http-ztkwidvb,GenerateName:,Namespace:serving-tests,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,},
              Spec: (v1alpha1.ConfigurationSpec) {
               DeprecatedGeneration: (int64) 0,
               Build: (*v1alpha1.RawExtension)(<nil>),
               RevisionTemplate: (v1alpha1.RevisionTemplateSpec) &ObjectMeta{Name:,GenerateName:,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,}
              },
              Status: (v1alpha1.ConfigurationStatus) {
               Status: (v1alpha1.Status) {
                ObservedGeneration: (int64) 0,
                Conditions: (v1alpha1.Conditions) <nil>
               },
               ConfigurationStatusFields: (v1alpha1.ConfigurationStatusFields) {
                LatestReadyRevisionName: (string) "",
                LatestCreatedRevisionName: (string) ""
               }
              }
             }),
             Service: (*v1alpha1.Service)(<nil>),
             Revision: (*v1alpha1.Revision)(<nil>)
            }
        errorcondition_test.go:166: When the containers keep crashing, the Configuration should have error status.
        errorcondition_test.go:182: Failed to validate configuration state: %!s(<nil>)
    --- FAIL: TestContainerExitingMsg/tcp (18.28s)
        errorcondition_test.go:157: Creating a new Configuration failing
        util.go:34: resource (test.ResourceObjects) {
             Route: (*v1alpha1.Route)(<nil>),
             Config: (*v1alpha1.Configuration)(0xc0005ce400)({
              TypeMeta: (v1.TypeMeta) &TypeMeta{Kind:,APIVersion:,},
              ObjectMeta: (v1.ObjectMeta) &ObjectMeta{Name:container-exiting-msg-tcp-eznuajeg,GenerateName:,Namespace:serving-tests,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,},
              Spec: (v1alpha1.ConfigurationSpec) {
               DeprecatedGeneration: (int64) 0,
               Build: (*v1alpha1.RawExtension)(<nil>),
               RevisionTemplate: (v1alpha1.RevisionTemplateSpec) &ObjectMeta{Name:,GenerateName:,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,}
              },
              Status: (v1alpha1.ConfigurationStatus) {
               Status: (v1alpha1.Status) {
                ObservedGeneration: (int64) 0,
                Conditions: (v1alpha1.Conditions) <nil>
               },
               ConfigurationStatusFields: (v1alpha1.ConfigurationStatusFields) {
                LatestReadyRevisionName: (string) "",
                LatestCreatedRevisionName: (string) ""
               }
              }
             }),
             Service: (*v1alpha1.Service)(<nil>),
             Revision: (*v1alpha1.Revision)(<nil>)
            }
        errorcondition_test.go:166: When the containers keep crashing, the Configuration should have error status.
        errorcondition_test.go:190: When the containers keep crashing, the revision should have error status.
        errorcondition_test.go:204: Failed to validate revision state: %!s(<nil>)
FAIL
```
</details>

wtam2018 replied:
I doubt that tcp/http "sub-tests" will ever run in parallel within the same test unit.   So, I'd keep the `t.Parallel()` at line 120 (at the beginning of the function) and delete the `t.Parallel()` at line 151 (inside the for-loop).
As far as the log output, please try increasing the count and check the log output.  e.g. ```go test -v -tags=e2e -count=8 -cpu=8 ./test/conformance -run ^TestContainerExitingMsg$ --timeout 30m```.    I agree that it is not easy to cause the log to mix up as output is printed when a test is complete.

yt3liu replied:
I think the subtests are run in parallel within the test unit. 

For the current test i.e. `t.Parallel()` within each subtests, the output is 
```
=== RUN   TestContainerExitingMsg
=== RUN   TestContainerExitingMsg/http
=== PAUSE TestContainerExitingMsg/http
=== RUN   TestContainerExitingMsg/tcp
=== PAUSE TestContainerExitingMsg/tcp
=== CONT  TestContainerExitingMsg/http
=== CONT  TestContainerExitingMsg/tcp
<A single wait before it prints the test result>
...
```

After moving `t.Parallel()` to the beginning

```
=== RUN   TestContainerExitingMsg
=== PAUSE TestContainerExitingMsg
=== CONT  TestContainerExitingMsg
=== RUN   TestContainerExitingMsg/http 
<There's a wait before it prints the next line> 
=== RUN   TestContainerExitingMsg/tcp
<Another wait before it prints the test result>
...
```

<details>
  <summary>Full output when `t.Parallel()` is moved to the beginning of the test</summary>

```
=== RUN   TestContainerExitingMsg
=== PAUSE TestContainerExitingMsg
=== CONT  TestContainerExitingMsg
=== RUN   TestContainerExitingMsg/http
=== RUN   TestContainerExitingMsg/tcp
--- PASS: TestContainerExitingMsg (164.48s)
    --- PASS: TestContainerExitingMsg/http (122.33s)
        errorcondition_test.go:158: Creating a new Configuration failing
        util.go:34: resource (test.ResourceObjects) {
             Route: (*v1alpha1.Route)(<nil>),
             Config: (*v1alpha1.Configuration)(0xc000073400)({
              TypeMeta: (v1.TypeMeta) &TypeMeta{Kind:,APIVersion:,},
              ObjectMeta: (v1.ObjectMeta) &ObjectMeta{Name:container-exiting-msg-http-smorkmba,GenerateName:,Namespace:serving-tests,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,},
              Spec: (v1alpha1.ConfigurationSpec) {
               DeprecatedGeneration: (int64) 0,
               Build: (*v1alpha1.RawExtension)(<nil>),
               RevisionTemplate: (v1alpha1.RevisionTemplateSpec) &ObjectMeta{Name:,GenerateName:,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,}
              },
              Status: (v1alpha1.ConfigurationStatus) {
               Status: (v1alpha1.Status) {
                ObservedGeneration: (int64) 0,
                Conditions: (v1alpha1.Conditions) <nil>
               },
               ConfigurationStatusFields: (v1alpha1.ConfigurationStatusFields) {
                LatestReadyRevisionName: (string) "",
                LatestCreatedRevisionName: (string) ""
               }
              }
             }),
             Service: (*v1alpha1.Service)(<nil>),
             Revision: (*v1alpha1.Revision)(<nil>)
            }
        errorcondition_test.go:166: When the containers keep crashing, the Configuration should have error status.
        errorcondition_test.go:190: When the containers keep crashing, the revision should have error status.
        errorcondition_test.go:207: When the revision has error condition, logUrl should be populated.
    --- PASS: TestContainerExitingMsg/tcp (42.16s)
        errorcondition_test.go:158: Creating a new Configuration failing
        util.go:34: resource (test.ResourceObjects) {
             Route: (*v1alpha1.Route)(<nil>),
             Config: (*v1alpha1.Configuration)(0xc00054e400)({
              TypeMeta: (v1.TypeMeta) &TypeMeta{Kind:,APIVersion:,},
              ObjectMeta: (v1.ObjectMeta) &ObjectMeta{Name:container-exiting-msg-tcp-cybbzcpn,GenerateName:,Namespace:serving-tests,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,},
              Spec: (v1alpha1.ConfigurationSpec) {
               DeprecatedGeneration: (int64) 0,
               Build: (*v1alpha1.RawExtension)(<nil>),
               RevisionTemplate: (v1alpha1.RevisionTemplateSpec) &ObjectMeta{Name:,GenerateName:,Namespace:,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,}
              },
              Status: (v1alpha1.ConfigurationStatus) {
               Status: (v1alpha1.Status) {
                ObservedGeneration: (int64) 0,
                Conditions: (v1alpha1.Conditions) <nil>
               },
               ConfigurationStatusFields: (v1alpha1.ConfigurationStatusFields) {
                LatestReadyRevisionName: (string) "",
                LatestCreatedRevisionName: (string) ""
               }
              }
             }),
             Service: (*v1alpha1.Service)(<nil>),
             Revision: (*v1alpha1.Revision)(<nil>)
            }
        errorcondition_test.go:166: When the containers keep crashing, the Configuration should have error status.
        errorcondition_test.go:190: When the containers keep crashing, the revision should have error status.
        errorcondition_test.go:207: When the revision has error condition, logUrl should be populated.
PASS
```
</details>

Thanks for suggesting a way to see if the log output can be mixed up. I still running `go test -v -tags=e2e -count=8 -cpu=8 ./test/conformance -run ^TestContainerExitingMsg$ --timeout 30m` in the background to see if I can at least see a case where the log output is mixed up, but I haven't seen the case where the log is mixed up. Since it is not easy to get to this state, I think the current logging is enough to determine which of the http/tcp case fails.

wtam2018 replied:
Yes. Subtests do indeed run in parallel.

yt3liu commented:
> I don't think this is the right test for these to be added. The readinessProbe is only here to make the test deterministic. We should have a separate test that tests that readiness/liveness Probes are "delayed" until the pod is really up (which this test won't check).

Could you please elaborate on what the readinessProbe is doing in the test `TestContainerExitingMsg`? The test is currently running against `failing` test image. From the comment, (https://github.com/knative/serving/blob/master/test/test_images/failing/failing.go#L28-L30)
```
        // Sleep for 10 seconds to force a race condition, where this
	// container becomes ready if no readinessProbe is set.
	time.Sleep(10 * time.Second)
```
1. In this case, doesn't the test `TestContainerExitingMsg` testing the exiting message when readiness probe is set? Could you please point out what I misunderstood here? 
1. What is the "delayed" you are referring to in the statement "We should have a separate test that tests that readiness/liveness Probes are "delayed" until the pod is really up (which this test won't check)."

markusthoemmes commented:
@yt3liu The the image in the test is failing anyway I think (doesn't it exit with an error code?). The readinessProbe is only there so the container doesn't go into a ready state intermittently (before the container exits).

Thinking about it, you might be right adding a livenessProbe here. Maybe we should change the image though to not actually crash the container, to make sure the livenessProbe actually checks something?

yt3liu commented:
> @yt3liu The the image in the test is failing anyway I think (doesn't it exit with an error code?). The readinessProbe is only there so the container doesn't go into a ready state intermittently (before the container exits).

Sorry, I am still not sure why the current `TestContainerExitingMsg` does not test the readinessProbe. I understand the readiness probe is there to prevent the container to go into the ready state intermittently, but doesn't that mean it's testing the readiness probe that it's checking for the readiness of the container?

If we only want to test the container exiting message for this test case, what is the reason for adding the extra delay in the test image and readiness probe? I removed the delay in the test image and removed the readiness probe from the test configuration and ran the test for 10+ times. I haven't seen a case where the test fails. Am I unlucky where it's passing for me all the time? What is the value of adding this extra 10 seconds sleep in the failing test image? 

> Thinking about it, you might be right adding a livenessProbe here. Maybe we should change the image though to not actually crash the container, to make sure the livenessProbe actually checks something?

@markusthoemmes I actually have another PR https://github.com/knative/serving/pull/3524 for removing the liveness probe check in the runtime contract. I would really appreciate it if you can also chime in your thoughts on the liveness/readiness probe there.

markusthoemmes commented:
> If we only want to test the container exiting message for this test case, what is the reason for adding the extra delay in the test image and readiness probe? I removed the delay in the test image and removed the readiness probe from the test configuration and ran the test for 10+ times. I haven't seen a case where the test fails. Am I unlucky where it's passing for me all the time? What is the value of adding this extra 10 seconds sleep in the failing test image?

Yes, that's sheer luck. I've seen it flake around on the CI system at least. I added the sleep to solidify the need for the readinessProbe so nobody removes it accidentally.

I might've been thinking along the wrong lines the whole time here. The adjustments are fine as-is.

wtam2018 commented:
/lgtm

dgerd commented on test/conformance/errorcondition_test.go:
```diff
@@ -117,80 +117,97 @@ func TestContainerErrorMsg(t *testing.T) {
-				return true, nil
+			names := test.ResourceNames{
+				Config: test.ObjectNameForTest(t),
+				Image:  "failing",
```
There are image constants in https://github.com/knative/serving/blob/master/test/conformance/util.go#L42 

I would add this here and reference the constant.

yt3liu replied:
Updated to use constants.

dgerd commented on test/conformance/errorcondition_test.go:
```diff
@@ -117,80 +117,97 @@ func TestContainerErrorMsg(t *testing.T) {
+		Name: "http",
+		ReadinessProbe: &corev1.Probe{
+			Handler: corev1.Handler{
+				HTTPGet: &corev1.HTTPGetAction{},
```
Is a "Path" not required here?

yt3liu replied:
The path is optional. Since the image is failing, I do not think it's providing any extra context by adding the path since nothing is exposed in the failing image.  https://github.com/knative/serving/blob/master/vendor/k8s.io/api/core/v1/types.go#L1878-L1880

dgerd commented on test/conformance/errorcondition_test.go:
```diff
@@ -118,79 +118,97 @@ func TestContainerErrorMsg(t *testing.T) {
+			if _, err := test.CreateConfiguration(t, clients, names, &test.Options{ReadinessProbe: tt.ReadinessProbe}); err != nil {
+				t.Fatalf("Failed to create configuration %s: %v", names.Config, err)
+			}
+			defer test.TearDown(clients, names)
```
Move this and the line below up above the CreateConfiguration otherwise this can leak.

yt3liu replied:
Thanks for pointing this out. Moved it up to immediately after clients, names creation.

dgerd commented:
/lgtm
/approve

dgerd commented:
/lgtm
